### DISCUSSION TITLE
"COLUMNS" in "curl --help all"

### DISCUSSION THREAD

**Discussion by gvanem** [curl/curl#13141] [Ideas]
As `$COLUMNS` is checked in the progress-bar, could it also be checked and used in `curl -help all`?

Now a `curl -help all | head -n 20` (with no `set COLUMNS=120` under Windows) it looks like this:
![image](https://github.com/curl/curl/assets/945271/449ccc50-3f49-419f-bf9e-37a63fd79a9a)

With a simple patch to `src/tool_help.c` (see below) and a `set COLUMNS=120` , it could look like this:
![image](https://github.com/curl/curl/assets/945271/23f33337-a0c2-4689-9984-309c20a4e078)

Which looks best? The patch I've used:
```diff
--- a/src/tool_help.c 2023-12-16 14:22:16
+++ b/src/tool_help.c 2024-03-17 16:52:03
@@ -80,6 +80,18 @@
   size_t longopt = 5;
   size_t longdesc = 5;

+  static curl_off_t columns = -1;
+
+  if (columns == -1) {
+    char *end, *env = curlx_getenv("COLUMNS");
+
+    columns = 80;
+    if (env) {
+      curlx_strtoofft(env, &end, 10, &columns);
+      curl_free(env);
+    }
+  }
+
   for(i = 0; helptext[i].opt; ++i) {
     size_t len;
     if(!(helptext[i].categories & category))
@@ -91,8 +103,8 @@
     if(len > longdesc)
       longdesc = len;
   }
-  if(longopt + longdesc > 80)
-    longopt = 80 - longdesc;
+  if(longopt + longdesc > (size_t)columns)
+    longopt = (size_t)columns - longdesc;

   for(i = 0; helptext[i].opt; ++i)
     if(helptext[i].categories & category) {
```

**Comment by bagder** [comment]
Sure why not!

But also: we should extract the "find the number of columns" logic from `progressbarinit()` and use instead of simply using `COLUMNS` because that variable is not set all the times and there are better ways to find the terminal width.

**Reply by bagder** [reply]
See  #13171
